Add blocking function fetch logic

lisajian opened the pull request:
### Description

Adds logic to fetch configured blocking functions and perform fetch on impacted endpoints. 

Corresponding internal bug: b/192387243

### Scenarios Tested

`npm test` passes

### Sample Commands

N/A

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -225,14 +228,29 @@ function signUp(
+        state,
+        BlockingFunctionEvents.BEFORE_SIGN_IN,
+        updates
+      );
```
```suggestion
      ).extraClaims;
```

yuchenshi replied:
nvm this works but just the variable name is misleading

lisajian commented on src/emulator/auth/operations.ts:
```diff
@@ -2830,6 +2904,246 @@ function updateTenant(
 }
 
+// TODO: Timeout is 60s. Should we make the timeout an emulator configuration?
+async function fetchBlockingFunctionWithTimeoutAndUpdateUser(
```
- Split logic for fetching from logic that applies modifiable fields
- Rename to something shorter, e.g. `fetchBlockingFunction`
- Generally will want to return update instead of having side effect

yuchenshi reviewed (changes requested):
BLOCKING: I think passing `updates` to `fetchBlockingFunction` is too fragile especially given that `updates` are only a tiny fraction of touched fields on most sign-ins. Let's talk offline to brainstorm a better high-level design.

We also need some tests for what was passed *into* the blocking function (i.e. request body) -- that would have caught some issues caused by the statement above.

yuchenshi commented:
Please kindly re-request review when this is ready for another round of review.

lisajian commented on src/emulator/auth/state.ts:
```diff
@@ -345,6 +353,16 @@ export abstract class ProjectState {
     return this.getUserByLocalIdAssertingExists(localId);
   }
 
+  getAllPhoneNumbersByLocalId(localId: string): string[] {
```
Ohh, good call, thanks!

lisajian commented on src/emulator/auth/state.ts:
```diff
@@ -57,6 +57,14 @@ export abstract class ProjectState {
 
   abstract get enableEmailLinkSignin(): boolean;
 
+  abstract get includeAccessToken(): boolean;
```
That makes sense, updated!

lisajian commented:
Newest revision includes the following changes:
- Modified impacted endpoints to call blocking functions for new/existing users
- Refactored endpoints (e.g. signInWithEmailLink and signInWithPhoneNumber) so it's easier to see when there's a new vs existing user
- Changed `fetchBlockingFunction()` so it takes a user instead of an update and only returns the updated fields instead of modifying `UserInfo` in place
- Changed `BeforeCreate` calls so the requests contains a pending `localId` and as much of the `UserInfo` available as possible prior to writing to store
- Changed `BeforeSignIn` calls so requests contains the `UserInfo` object with any updates from `BeforeCreate` updates

Hopefully I addressed everything, but let me know if I'm missing anything! Thanks for taking a look

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -2830,6 +2920,254 @@ function updateTenant(
+
+    for (const field of fields) {
+      if (
+        Object.values(BlockingFunctionModifiableFields).includes(
```
Blocking functions are user code and they can return anything they like. For example, a developer may be using plain JavaScript without type checking.

lisajian replied:
Ah true, I'm not sure what the backend behavior would be but I added a bit of logic to coerce both displayName and photoUrl to string and similarly for the booleans

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -223,12 +227,39 @@ function signUp(
+
+    if (reqBody.email) {
+      if (!user.disabled) {
+        const blockingResponse = await fetchBlockingFunction(
```
I believe the current logic calls the `beforeSignIn` blocking function even for an admin request (e.g. from AdQmin SDKs or Emulator UI). Is this the right behavior?

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -1393,23 +1462,23 @@ function signInWithEmailLink(
-    (state.mfaConfig.state === "ENABLED" || state.mfaConfig.state === "MANDATORY") &&
-    user.mfaInfo?.length
-  ) {
+  if (isMfaEnabled(state, user)) {
```
For disabled users with MFA, should the API also return an error here?

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -2800,6 +2994,276 @@ function updateTenant(
+    const text = await res.text();
+    assert(
+      res.ok,
+      `BLOCKING_FUNCTION_ERROR_RESPONSE: ((HTTP request to ${url} returned an error: ${text}))`
```
```suggestion
      `BLOCKING_FUNCTION_ERROR_RESPONSE: ((HTTP request to ${url} returned HTTP error ${res.status}: ${text}))`
```

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -2800,6 +2994,276 @@ function updateTenant(
+        case "photoUrl":
+        case "disabled":
+        case "emailVerified":
+          (updates as any)[field] = userRecord[field];
```
Is `as any` still needed? If so, maybe try `updates[field as keyof BlockingFunctionUpdates]`

lisajian replied:
Sick, updated!

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -2800,6 +2994,276 @@ function updateTenant(
+          updates.customAttributes = userRecord.customClaims;
+          break;
+        // Session claims are only returned in beforeSignIn. For more info,
+        // see https://cloud.google.com/identity-platform/docs/blocking-functions#modifying_a_user
```
If a developer makes a mistake and returns `sessionClaims` from `beforeCreate`, the implementation will try to parse it instead of ignores it. Is there anyway we can address this?

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -2830,6 +2920,254 @@ function updateTenant(
+  if (user.localId) {
+    const allPhoneNumbers = state.getAllPhoneNumbersByLocalId(user.localId);
+    for (const phoneNumber of allPhoneNumbers) {
+      const provider: Provider = {
```
I think the Auth Emulator implementation is funkier than the internal link -- since the Auth Emulator tries to keep `user.providerUserInfo` synced with phone numbers when writing to DB, Line 3191 will transform that entry into `provider_data`. The logic here will probably add yet another entry to `provider_data`, creating a duplicate. Feel free to add a test to prove me wrong.

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -2800,6 +2994,276 @@ function updateTenant(
+      : [];
+    for (const phoneNumber of allPhoneNumbers) {
+      const provider: Provider = {
+        provider_id: PROVIDER_PASSWORD,
```
```suggestion
        provider_id: PROVIDER_PHONE,
```

yuchenshi commented on src/emulator/auth/operations.ts:
```diff
@@ -2830,6 +2920,254 @@ function updateTenant(
+    }
+  }
+  if (user.email && (user.passwordHash || user.emailLinkSignin)) {
+    const provider: Provider = {
```
Same concern as above

yuchenshi commented on src/emulator/auth/state.ts:
```diff
@@ -53,14 +54,21 @@ export abstract class ProjectState {
-        return user;
+      if (!this.users.has(localId) && !this.pendingLocalIds.has(localId)) {
+        // Create a pending localId until user is created
+        this.pendingLocalIds.add(localId);
```
This creates a memory leak if the blocking functions throws and the `localId` is never used. (Theoretically, if this happens enough times we will even run out of IDs.) I think this is probably acceptable, but worth a comment to document the tradeoff.

yuchenshi reviewed (commented):
Partial LGTM. I have not reviewed the tests line by line yet, but I believe we're still lacking coverage for blocking functions _request_ payloads. See comments inline.

yuchenshi commented on src/test/emulators/auth/mfa.spec.ts:
```diff
@@ -462,4 +469,210 @@ describeAuthEmulator("mfa enrollment", ({ authApi, getClock }) => {
+      nock.cleanAll();
+    });
+
+    it("mfaSignIn:finalize should update modifiable fields before sign in", async () => {
```
nit: add a test that blocking function is not called if MFA is present on user but *not* performed, and `displayName` is not updated.

yuchenshi commented on src/test/emulators/auth/mfa.spec.ts:
```diff
@@ -462,4 +469,210 @@ describeAuthEmulator("mfa enrollment", ({ authApi, getClock }) => {
+          expectStatusCode(200, res);
+          expect(res.body).not.to.have.property("idToken");
+          expect(res.body).not.to.have.property("refreshToken");
+          const mfaPendingCredential = res.body.mfaPendingCredential as string;
```
nit: extract a helper function for extracting mfaPendingCredential and mfaEnrollmentId from response (mainly to clean the code up visually, since the exact response format is not the focus of the test here)

yuchenshi commented on src/test/emulators/auth/signUp.spec.ts:
```diff
@@ -572,4 +581,273 @@ describeAuthEmulator("accounts:signUp", ({ authApi }) => {
+        });
+    });
+
+    it("should disable email/password user if set", async () => {
```
```suggestion
    it("should disable new user if set", async () => {
```

yuchenshi commented on src/test/emulators/auth/signUp.spec.ts:
```diff
@@ -572,4 +581,273 @@ describeAuthEmulator("accounts:signUp", ({ authApi }) => {
+
+      // Shouldn't trigger nock calls
+      expect(nock.isDone()).to.be.false;
+      nock.cleanAll();
```
This should be in a `afterEach` block so that the cleanup still happens even if the test fails somehow

yuchenshi reviewed (approved):
LGTM! Please add at least one test for the blocking function request payload though (i.e. the `jwt` that is sent to the blocking function). This reduces chances of regression in the future.

lisajian commented:
> LGTM! Please add at least one test for the blocking function request payload though (i.e. the `jwt` that is sent to the blocking function). This reduces chances of regression in the future.

Added to signInWithEmailLink
